File: version/v1_19_x/src/main/java/org/aresclient/ares/impl/minecraft/render/ScreenAdapter.java
package org.aresclient.ares.impl.minecraft.render;

import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.Text;
import org.aresclient.ares.api.minecraft.render.Screen;

public class ScreenAdapter extends net.minecraft.client.gui.screen.Screen {
    private final Screen screen;

    public ScreenAdapter(Screen screen) {
        super(Text.literal(screen.getTitle()));
        RenderSystem.assertOnRenderThreadOrInit();
        this.screen = screen;
    }

    @Override
    public void init() {
        screen.resize(width, height);
        screen.update();
    }

    @Override
    public void render(MatrixStack matrixStack, int mouseX, int mouseY, float partialTicks) {
        screen.render(mouseX, mouseY, partialTicks);
        super.render(matrixStack, mouseX, mouseY, partialTicks);
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int mouseButton) {
        screen.click((int) mouseX, (int) mouseY, mouseButton);
        return super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    public boolean mouseReleased(double mouseX, double mouseY, int mouseButton) {
        screen.release((int) mouseX, (int) mouseY, mouseButton);
        return super.mouseReleased(mouseX, mouseY, mouseButton);
    }

    @Override
    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
        screen.type(null, keyCode);
        return super.keyPressed(keyCode, scanCode, modifiers);
    }

    @Override
    public boolean charTyped(char character, int keycode) {
        screen.type(character, keycode);
        return super.charTyped(character, keycode);
    }

    @Override
    public boolean mouseScrolled(double mouseX, double mouseY, double value) {
        screen.scroll((int) mouseX, (int) mouseY, value * 15);
        return super.mouseScrolled(mouseX, mouseY, value);
    }

    @Override
    public boolean shouldPause() {
        return screen.shouldPause();
    }

    @Override
    public boolean shouldCloseOnEsc() {
        return screen.shouldCloseOnEsc();
    }

    @Override
    public void close() {
        screen.close();
        super.close();
    }
}

File: version/v1_19_x/src/main/kotlin/org/aresclient/ares/impl/ScreenAdapter.kt
package org.aresclient.ares.impl

import net.minecraft.client.MinecraftClient
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.LiteralTextContent
import net.minecraft.text.MutableText
import org.aresclient.ares.api.ScreenContext
import org.aresclient.ares.renderer.Screen

class ScreenAdapter(private val screen: Screen): net.minecraft.client.gui.screen.Screen(MutableText.of(LiteralTextContent(screen.title))) {
    init {
        screen.setContext(this as ScreenContext)
    }

    override fun init() {
        screen.init()
    }

    override fun render(matrixStack: MatrixStack?, mouseX: Int, mouseY: Int, partialTicks: Float) {
        screen.render(mouseX, mouseY, partialTicks)
        super.render(matrixStack, mouseX, mouseY, partialTicks)
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, mouseButton: Int): Boolean {
        screen.click(mouseX.toInt(), mouseY.toInt(), mouseButton)
        return super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override fun mouseReleased(mouseX: Double, mouseY: Double, mouseButton: Int): Boolean {
        screen.release(mouseX.toInt(), mouseY.toInt(), mouseButton)
        return super.mouseReleased(mouseX, mouseY, mouseButton)
    }

    override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
        screen.type(null, keyCode)
        return super.keyPressed(keyCode, scanCode, modifiers)
    }

    override fun charTyped(character: Char, keycode: Int): Boolean {
        screen.type(character, keycode)
        return super.charTyped(character, keycode)
    }

    override fun mouseScrolled(mouseX: Double, mouseY: Double, value: Double): Boolean {
        screen.scroll(mouseX.toInt(), mouseY.toInt(), value * 15)
        return super.mouseScrolled(mouseX, mouseY, value)
    }

    override fun resize(client: MinecraftClient?, width: Int, height: Int) {
        screen.resize(width, height)
        super.resize(client, width, height)
    }

    override fun shouldPause(): Boolean {
        return screen.shouldPause()
    }

    override fun shouldCloseOnEsc(): Boolean {
        return screen.shouldCloseOnEsc()
    }

    override fun close() {
        screen.close()
        super.close()
    }
}